libcxx/include/flat_set

```diff
@@ -0,0 +1,913 @@
+    : __compare_(__comp) {}
+
+  template <class _Alloc>
+    requires uses_allocator_v<container_type, _Alloc>
+  flat_set(const key_compare& __comp, const _Alloc& __a)
+    : __compare_(__comp), __c_(std::__make_obj_using_allocator<container_type>(__a)) {}
+
+  template <class _Alloc>
+    requires uses_allocator_v<container_type, _Alloc>
+  explicit flat_set(const _Alloc& __a)
+    : __c_(std::__make_obj_using_allocator<container_type>(__a)) {}
+
+  template <class _Alloc>
+    requires uses_allocator_v<container_type, _Alloc>
+  flat_set(const flat_set& __fs, const _Alloc& __a)
+    : __compare_(__fs.__compare_), __c_(std::__make_obj_using_allocator<container_type>(__a, __fs.__c_)) {}
+
+  template <class _Alloc>
+    requires uses_allocator_v<container_type, _Alloc>
+  flat_set(flat_set&& __fs, const _Alloc& __a)
+    : __compare_(__fs.__compare_), __c_(std::__make_obj_using_allocator<container_type>(__a, std::move(__fs.__c_))) {
+    __fs.__c_.clear();
+  }
+
+  template <class _InIt>
+    requires __is_cpp17_input_iterator<_InIt>::value
+  flat_set(_InIt __first, _InIt __last, const key_compare& __comp = key_compare())
+    : __compare_(__comp) {
+    insert(__first, __last);
+  }
+
+  template <class _InIt, class _Alloc>
+    requires __is_cpp17_input_iterator<_InIt>::value && uses_allocator_v<container_type, _Alloc>
+  flat_set(_InIt __first, _InIt __last, const key_compare& __comp, const _Alloc& __a)
+    : __compare_(__comp), __c_(std::__make_obj_using_allocator<container_type>(__a)) {
+    insert(__first, __last);
+  }
+
+  template <class _InIt, class _Alloc>
+    requires __is_cpp17_input_iterator<_InIt>::value && uses_allocator_v<container_type, _Alloc>
+  flat_set(_InIt __first, _InIt __last, const _Alloc& __a)
+    : __c_(std::__make_obj_using_allocator<container_type>(__a)) {
+    insert(__first, __last);
+  }
+
+  template <__container_compatible_range<value_type> _Rp>
+  flat_set(from_range_t __fr, _Rp&& __r)
+    : flat_set(__fr, std::forward<_Rp>(__r), key_compare()) {}
+
+  template <__container_compatible_range<value_type> _Rp, class _Alloc>
+    requires uses_allocator_v<container_type, _Alloc>
+  flat_set(from_range_t __fr, _Rp&& __r, const _Alloc& __a)
+    : flat_set(__fr, std::forward<_Rp>(__r), key_compare(), __a) {}
+
+  template <__container_compatible_range<value_type> _Rp>
+  flat_set(from_range_t, _Rp&& __r, const key_compare& __comp)
+    : flat_set(__comp) {
+    insert_range(std::forward<_Rp>(__r));
+  }
+
+  template <__container_compatible_range<value_type> _Rp, class _Alloc>
+    requires uses_allocator_v<container_type, _Alloc>
+  flat_set(from_range_t, _Rp&& __r, const key_compare& __comp, const _Alloc& __a)
+    : flat_set(__comp, __a) {
+    insert_range(std::forward<_Rp>(__r));
+  }
+
+  template <class _InIt>
+  flat_set(sorted_unique_t, _InIt __first, _InIt __last, const key_compare& __comp = key_compare())
+    : __compare_(__comp), __c_(__first, __last) {}
```

Comment:
I think that would be a nice place to add a `_LIBCPP_DEBUG_ASSERT(is_sorted)`. In fact, please make sure that you have a `_LIBCPP_ASSERT` for each precondition in this class. When the preconditions are expensive to check, we'll hide them behind the debug mode.

Reply:
Oh, and let's add a libc++ specific test for all of those, they are easy to write.